### PR TITLE
Migrate CheckServiceAffinity custom predicate to Filter plugin

### LINKED ISSUE #84455: Migrate CheckServiceAffinity custom predicate to Filter plugin
**What would you like to be added**:
convert CheckServiceAffinity to a filter plugin

**Why is this needed**:
part of #84187 which is part of #83554

**Additional Notes**
Please refer to #84297 as an example.

/sig scheduling
/cc @ahg-g
/priority important-soon
/help

### PR DESCRIPTION
**Pull request by danielqsj**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Part of scheduler framework migration.

**Which issue(s) this PR fixes**:

Fixes #84455

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by ahg-g** [comment]
Thanks, we need to make a slight modification to how we handle custom predicates/priorities, @liu-cong will send an update soon

/hold

**Comment by ahg-g** [comment]
Hi Daniel, are you able to finish this PR? sorry for pushing, but the freeze deadline is very soon.

File: pkg/scheduler/algorithm_factory.go [hunk 1]
```diff
@@ -272,6 +273,10 @@ func RegisterCustomFitPredicate(policy schedulerapi.PredicatePolicy, args *plugi
 	// generate the predicate function, if a custom type is requested
 	if policy.Argument != nil {
 		if policy.Argument.ServiceAffinity != nil {
+			// map LabelsPresence policy to ConfigProducerArgs that's used to configure the ServiceAffinity plugin.
+			args.ServiceAffinityArgs = &serviceaffinity.Args{
+				Labels: policy.Argument.LabelsPresence.Labels,
+			}
```
[the remark this answers is not in the document]
Comment (danielqsj) [reply pkg/scheduler/algorithm_factory.go]: thanks, done

File: pkg/scheduler/algorithm_factory.go [hunk 2]
```diff
@@ -284,6 +289,8 @@ func RegisterCustomFitPredicate(policy schedulerapi.PredicatePolicy, args *plugi
 				predicates.RegisterPredicateMetadataProducer(policy.Name, precomputationFunction)
 				return predicate
 			}
+			// We do not allow specifying the name for custom plugins, see #83472
```
[the remark this answers is not in the document]
Comment (danielqsj) [reply pkg/scheduler/algorithm_factory.go]: thanks, done

File: pkg/scheduler/factory_test.go [hunk 3]
```diff
@@ -26,6 +26,8 @@ import (
 	"testing"
 	"time"
 
+	"k8s.io/kubernetes/pkg/scheduler/framework/plugins/serviceaffinity"
```
[the remark this answers is not in the document]
Comment (danielqsj) [reply pkg/scheduler/factory_test.go]: nice catch, done

**Comment by danielqsj** [comment]
@ahg-g PR updated, thanks for your reminder. :)

File: pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+		return nil, err
+	}
+	informerFactory := handle.SharedInformerFactory()
+	nodeInfoLister := handle.SnapshotSharedLister().NodeInfos()
+	podLister := handle.SnapshotSharedLister().Pods()
+	serviceLister := informerFactory.Core().V1().Services().Lister()
+	fitPredicate, predicateMetadataProducer := predicates.NewServiceAffinityPredicate(nodeInfoLister, podLister, serviceLister, args.Labels)
+
+	// Once we generate the predicate we should also Register the Precomputation
+	predicates.RegisterPredicateMetadataProducer("ServiceAffinityMetaProducer", predicateMetadataProducer)
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go]: It is already registered here: https://github.com/kubernetes/kubernetes/pull/84635/files?file-filters%5B%5D=.go#diff-181c3e9bbe3177aa8bf9744daac6eb49R293

Using a different name for registration will cause it to register twice. I assume you wanted to register here for the tests, we can make this call idempotent by making sure we use the same name in both places: ```Name```

File: pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	framework "k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1"
+	"k8s.io/kubernetes/pkg/scheduler/nodeinfo"
+)
+
+// Name of this plugin.
+const Name = "ServiceAffinity"
+
+// Args holds the args that are used to configure the plugin.
+type Args struct {
+	// The list of labels that identify node "groups"
```
Comment (liu-cong) [inline pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go]: The comment is wrong. I think it's taken from the NodeLabel predicate.

File: pkg/scheduler/algorithm_factory.go [hunk 6]
```diff
@@ -273,32 +274,43 @@ func RegisterCustomFitPredicate(policy schedulerapi.PredicatePolicy, args *plugi
...
 				predicate, precomputationFunction := predicates.NewServiceAffinityPredicate(
 					args.NodeInfoLister,
 					args.PodLister,
 					args.ServiceLister,
-					policy.Argument.ServiceAffinity.Labels,
+					pluginArgs.ServiceAffinityArgs.Labels,
 				)
 
 				// Once we generate the predicate we should also Register the Precomputation
 				predicates.RegisterPredicateMetadataProducer(policy.Name, precomputationFunction)
```
Comment (ahg-g) [inline pkg/scheduler/algorithm_factory.go]: use ```policyName```

File: pkg/scheduler/algorithm_factory.go [hunk 7]
```diff
@@ -263,7 +264,7 @@ func RegisterFitPredicateFactory(name string, predicateFactory FitPredicateFacto
 
 // RegisterCustomFitPredicate registers a custom fit predicate with the algorithm registry.
 // Returns the name, with which the predicate was registered.
-func RegisterCustomFitPredicate(policy schedulerapi.PredicatePolicy, args *plugins.ConfigProducerArgs) string {
+func RegisterCustomFitPredicate(policy schedulerapi.PredicatePolicy, pluginArgs *plugins.ConfigProducerArgs) string {
 	var predicateFactory FitPredicateFactory
 	var ok bool
 	name := policy.Name
```
Comment (ahg-g) [inline pkg/scheduler/algorithm_factory.go]: s/name/policyName

File: pkg/scheduler/algorithm_factory.go [hunk 8]
```diff
@@ -273,32 +274,43 @@ func RegisterCustomFitPredicate(policy schedulerapi.PredicatePolicy, args *plugi
...
 					args.ServiceLister,
-					policy.Argument.ServiceAffinity.Labels,
+					pluginArgs.ServiceAffinityArgs.Labels,
 				)
 
 				// Once we generate the predicate we should also Register the Precomputation
 				predicates.RegisterPredicateMetadataProducer(policy.Name, precomputationFunction)
 				return predicate
 			}
+			// We use the ServiceAffinity plugin name for all ServiceAffinity custom priorities.
```
Comment (ahg-g) [inline pkg/scheduler/algorithm_factory.go]: move this to the beginning of the if statement ```if policy.Argument.ServiceAffinity != nil```

File: pkg/scheduler/algorithm_factory.go [hunk 9]
```diff
@@ -273,32 +274,43 @@ func RegisterCustomFitPredicate(policy schedulerapi.PredicatePolicy, args *plugi
...
 			}
 			predicateFactory = func(_ PluginFactoryArgs) predicates.FitPredicate {
 				return predicates.NewNodeLabelPredicate(
-					args.NodeLabelArgs.PresentLabels,
-					args.NodeLabelArgs.AbsentLabels,
+					pluginArgs.NodeLabelArgs.PresentLabels,
+					pluginArgs.NodeLabelArgs.AbsentLabels,
 				)
 			}
 			// We use the NodeLabel plugin name for all NodeLabel custom priorities.
```
Comment (ahg-g) [inline pkg/scheduler/algorithm_factory.go]: ditto, please move this to the beginning of the if statement. and change lines 323 and 328 to use policyName.

**Review by ahg-g — COMMENTED** [review]
thanks!

**Comment by ahg-g** [comment]
> **What type of PR is this?**
> /kind feature
> 
> **What this PR does / why we need it**:
> Part of scheduler framework migration.
> 
> **Which issue(s) this PR fixes**:
> 
> Fixes #84455
> 
> **Special notes for your reviewer**:
> 
> **Does this PR introduce a user-facing change?**:
> 
> ```
> User can now use component config to configure ServiceAffinity plugin for the scheduler framework.
> ```
> 
> **Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
> 
> ```
> 
> ```

also, please remove this user facing statement and just put NONE.

File: pkg/scheduler/factory_test.go [hunk 10]
```diff
@@ -137,6 +138,21 @@ func TestCreateFromConfig(t *testing.T) {
...
+	if _, ok := findPlugin(serviceaffinity.Name, "FilterPlugin", conf); !ok {
+		t.Fatalf("ServiceAffinity plugin not exist in framework.")
+	}
+	// Verify that the policy config is converted to plugin config for service affinity predicate.
+	serviceAffinityConfig := findPluginConfig(serviceaffinity.Name, conf)
+	encoding, err = json.Marshal(serviceAffinityConfig)
+	if err != nil {
+		t.Errorf("Failed to marshal %+v: %v", serviceAffinityConfig, err)
+	}
+	want = `{"Name":"ServiceAffinity","Args":{"labels":["zone"]}}`
```
Comment (liu-cong) [inline pkg/scheduler/factory_test.go]: Please update the test case to include multiple policy config for the ServiceAffinity and check that the labels are aggregated.

File: pkg/scheduler/algorithm_factory.go (hunk 7 above)
Comment (danielqsj) [reply pkg/scheduler/algorithm_factory.go]: done

File: pkg/scheduler/algorithm_factory.go (hunk 6 above)
Comment (danielqsj) [reply pkg/scheduler/algorithm_factory.go]: done

File: pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go (hunk 4 above)
Comment (danielqsj) [reply pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go]: Thanks, I miss this name. Fixed

File: pkg/scheduler/algorithm_factory.go (hunk 9 above)
Comment (danielqsj) [reply pkg/scheduler/algorithm_factory.go]: done

**Comment by danielqsj** [comment]
@ahg-g @liu-cong could you help check this test failure: https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/84635/pull-kubernetes-bazel-test/1192108607244603394 . 
It seems that the return of `ListPlugins` is not sorted. Right?

**Comment by danielqsj** [comment]
too late in China, I will keep working on this PR tomorrow morning.

**Comment by liu-cong** [comment]
> @ahg-g @liu-cong could you help check this test failure: https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/84635/pull-kubernetes-bazel-test/1192108607244603394 .
> It seems that the return of `ListPlugins` is not sorted. Right?

Not sure what order does it use. But if the order is fixed, I think you can adjust your expected list in the test.

**Comment by ahg-g** [comment]
> > @ahg-g @liu-cong could you help check this test failure: https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/84635/pull-kubernetes-bazel-test/1192108607244603394 .
> > It seems that the return of `ListPlugins` is not sorted. Right?
> 
> Not sure what order does it use. But if the order is fixed, I think you can adjust your expected list in the test.

Yes, the order if fixed, just adjust the test to the expected order as Cong suggested.

File: pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+// Name returns name of the plugin. It is used in logs, etc.
+func (pl *ServiceAffinity) Name() string {
+	return Name
+}
+
+// Filter invoked at the filter extension point.
+func (pl *ServiceAffinity) Filter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodeInfo *nodeinfo.NodeInfo) *framework.Status {
+	meta, ok := migration.PredicateMetadata(cycleState).(predicates.PredicateMetadata)
+	if !ok {
+		return migration.ErrorToFrameworkStatus(fmt.Errorf("%+v convert to predicates.PredicateMetadata error", cycleState))
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go]: cycleState can potentially be large, just do this:
```return framework.NewStatus(Error, "looking up PredicateMetadata")```

**Comment by ahg-g** [comment]
the compatibility test still failing, you need to place this plugin in the right order.

**Comment by danielqsj** [comment]
@ahg-g the test result is different in each try:
1. https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/84635/pull-kubernetes-bazel-test/1192119682958626818
2. https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/84635/pull-kubernetes-bazel-test/1192280352714919938

File: pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go (hunk 11 above)
Comment (danielqsj) [reply pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go]: sure

**Comment by danielqsj** [comment]
@ahg-g I resolve the test issues by sorting the plugin slice before compare. PTAL

**Comment by ahg-g** [comment]
> @ahg-g I resolve the test issues by sorting the plugin slice before compare. PTAL

The order for predicates is actually part of the test, we want to make sure we are respecting the [order specified here](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithm/predicates/predicates.go#L138). 

The reason we are getting a different order for custom predicates is because when we register them, we are using a different name from the one in the [order specified here](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithm/predicates/predicates.go#L138) list. When generating the plugin config, we are appending those at the end according to their order in a temporary set: https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/factory.go#L586

NodeLabel and ServiceAffinity should be registered as predicate using the names defined here https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithm/predicates/predicates.go#L53

File: pkg/scheduler/algorithm_factory.go [hunk 12]
```diff
@@ -263,60 +264,73 @@ func RegisterFitPredicateFactory(name string, predicateFactory FitPredicateFacto
...
 
 	validatePredicateOrDie(policy)
 
 	// generate the predicate function, if a custom type is requested
 	if policy.Argument != nil {
 		if policy.Argument.ServiceAffinity != nil {
+			// We use the ServiceAffinity plugin name for all ServiceAffinity custom priorities.
+			// It may get called multiple times but we essentially only register one instance of ServiceAffinity predicate.
+			// This name is then used to find the registered plugin and run the plugin instead of the predicate.
+			policyName = serviceaffinity.Name
```
Comment (ahg-g) [inline pkg/scheduler/algorithm_factory.go]: ```policyName = predicates.CheckServiceAffinityPred```

File: pkg/scheduler/framework/plugins/default_registry.go [hunk 13]
```diff
@@ -209,6 +213,12 @@ func NewDefaultConfigProducerRegistry() *ConfigProducerRegistry {
 			pluginConfig = append(pluginConfig, makePluginConfig(nodelabel.Name, args.NodeLabelArgs))
 			return
 		})
+	registry.RegisterPredicate(serviceaffinity.Name,
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/default_registry.go]: s/serviceaffinity.Name/predicates.CheckServiceAffinityPred

File: pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+		return nil, err
+	}
+	informerFactory := handle.SharedInformerFactory()
+	nodeInfoLister := handle.SnapshotSharedLister().NodeInfos()
+	podLister := handle.SnapshotSharedLister().Pods()
+	serviceLister := informerFactory.Core().V1().Services().Lister()
+	fitPredicate, predicateMetadataProducer := predicates.NewServiceAffinityPredicate(nodeInfoLister, podLister, serviceLister, args.Labels)
+
+	// Once we generate the predicate we should also Register the Precomputation
+	predicates.RegisterPredicateMetadataProducer(Name, predicateMetadataProducer)
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go]: s/Name/predicates.CheckServiceAffinityPred

File: pkg/scheduler/framework/plugins/default_registry.go (hunk 13 above)
Comment (ahg-g) [reply pkg/scheduler/framework/plugins/default_registry.go]: in line 210: s/nodelabel.Name/predicates.CheckNodeLabelPresencePred

File: pkg/scheduler/algorithm_factory.go [hunk 15]
```diff
@@ -263,60 +264,73 @@ func RegisterFitPredicateFactory(name string, predicateFactory FitPredicateFacto
...
 				// Once we generate the predicate we should also Register the Precomputation
-				predicates.RegisterPredicateMetadataProducer(policy.Name, precomputationFunction)
+				predicates.RegisterPredicateMetadataProducer(policyName, precomputationFunction)
 				return predicate
 			}
 		} else if policy.Argument.LabelsPresence != nil {
+			// We use the NodeLabel plugin name for all NodeLabel custom priorities.
+			// It may get called multiple times but we essentially only register one instance of NodeLabel predicate.
+			// This name is then used to find the registered plugin and run the plugin instead of the predicate.
+			policyName = nodelabel.Name
```
Comment (ahg-g) [inline pkg/scheduler/algorithm_factory.go]: ```policyName = predicates.CheckNodeLabelPresencePred```

File: pkg/scheduler/apis/config/testing/compatibility_test.go [hunk 16]
```diff
@@ -1333,6 +1322,10 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
 				seenPriorities.Insert(scoreToPriorityMap[p.Name])
 
 			}
+
+			sortPlugins(tc.wantPlugins)
+			sortPlugins(gotPlugins)
+
```
Comment (ahg-g) [inline pkg/scheduler/apis/config/testing/compatibility_test.go]: revert this.

File: pkg/scheduler/apis/config/testing/compatibility_test.go [hunk 17]
```diff
@@ -1372,3 +1365,14 @@ func pluginsToStringSet(plugins []config.Plugin) sets.String {
...
+func sortPlugins(allPlugins map[string][]config.Plugin) {
+	if allPlugins == nil {
+		return
+	}
+	for _, plugins := range allPlugins {
+		sort.Slice(plugins, func(i, j int) bool {
+			return plugins[i].Name < plugins[j].Name
+		})
+	}
+}
```
Comment (ahg-g) [inline pkg/scheduler/apis/config/testing/compatibility_test.go]: revert this.

**Comment by danielqsj** [comment]
@ahg-g Got it. Thanks for your guide.
PR updated. PTAL

**Comment by ahg-g** [comment]
look good, please squash.

**Comment by danielqsj** [comment]
sure, squashed. :)

**Comment by ahg-g** [comment]
/hold

can you please address Cong's comment related to the unit test.

File: pkg/scheduler/factory_test.go (hunk 10 above)
Comment (liu-cong) [reply pkg/scheduler/factory_test.go]: This is not addressed. We should update the tests.

Comment (liu-cong) [reply pkg/scheduler/factory_test.go]: Sorry, was looking at old version for some reason. 
/LGTM.

**Comment by liu-cong** [comment]
/hold cancel
My comments were addressed.

**Pull request merged**